nodejs/aws-infra/experimental/clusterAutoScaling.ts

```diff
@@ -0,0 +1,572 @@
+// Copyright 2016-2018, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import * as aws from "@pulumi/aws";
+import * as pulumi from "@pulumi/pulumi";
+
+import * as mod from ".";
+
+import { Overwrite, sha1hash } from "./../utils";
+
+export class ClusterAutoScalingLaunchConfiguration extends pulumi.ComponentResource {
+    public readonly instance: aws.ec2.LaunchConfiguration;
+
+    public readonly cluster: mod.Cluster;
+    /**
+     * Optional file system to mount.  Use [cluster.createFileSystem] to create an instance of this.
+     */
+    public readonly fileSystem?: mod.ClusterFileSystem;
+
+    public readonly instanceProfile: aws.iam.InstanceProfile;
+
+    /**
+     * Name to give the auto-scaling-group's cloudformation stack name.
+     */
+    public readonly stackName: pulumi.Output<string>;
+
+    constructor(name: string,
+                args: ClusterAutoScalingLaunchConfigurationArgs,
+                opts: pulumi.ComponentResourceOptions = {}) {
+        super("aws-infra.x.ClusterAutoScalingLaunchConfiguration", name, args, opts);
+
+        const parentOpts = { parent: this };
+
+        const cluster = args.cluster;
+
+        // Create the full name of our CloudFormation stack here explicitly. Since the CFN stack
+        // references the launch configuration and vice-versa, we use this to break the cycle.
+        // TODO[pulumi/pulumi#381]: Creating an S3 bucket is an inelegant way to get a durable,
+        // unique name.
+        const stackName = pulumi.output(args.stackName).apply(sn => sn || new aws.s3.Bucket(name, {}, parentOpts).id);
+
+        // Use the instance provided, or create a new one.
+        const instanceProfile = getInstanceProfile(name, args, parentOpts);
+        const fileSystem = args.fileSystem;
+
+        const securityGroups =
+            pulumi.output(args.securityGroups).apply(g => g || cluster.instanceSecurityGroups.map(g => g.id));
+        const instance = new aws.ec2.LaunchConfiguration(name, {
+            ...args,
+            securityGroups,
+            imageId: getEcsAmiId(args.ecsOptimizedAMIName),
+            instanceType: pulumi.output(args.instanceType).apply(t => t || "t2.micro"),
```

Comment:
Why not rely on the default assigned by the provider, if an explicit one is not specified?